Copy mutable variables on assignment (configuration_data and environment)

tp-m opened the pull request:
All assignments in meson should be by value, so mutable objects
(i.e. environment() and configuration_data()) should be copied
automatically on assignment.
    
Fixes #831

nirbheek commented:
Can we also get a test that does the same for the environment object? :)

nirbheek commented on mesonbuild/interpreter.py:
```diff
@@ -2460,6 +2456,9 @@ def assignment(self, node):
             raise InvalidCode('Tried to assign an invalid value to variable.')
+        # For mutable objects we need to make a copy on assignment
+        if isinstance(value, EnvironmentVariablesHolder) or isinstance(value, ConfigurationDataHolder):
+            value = copy.deepcopy(value)
```
I'm wondering if we should add a `mutable` flag on various variable holders and enforce that everywhere.

tp-m commented:
Instead of a flag I've added a MutableInterpreterObject class, which seemed the easiest since we have to check the instance type anyway if I'm not mistaken. (I might have misunderstood what you asked for though.)

I have added a test for the `environment()` object being copied properly as well.

nirbheek commented:
The new class actually makes more sense. LGTM.
